egauge/script/api_egauge.py

```diff
@@ -0,0 +1,148 @@
+#!./env/bin/python3
+"""
+This script makes egauge api requests using data retrieved from a database
+
+You can run this script using the form:
+python3 egauge_api.py <sensor id> <database url> <unit of time> <output file>
+
+Example command: python3 egauge_api.py 725 postgresql:///egauge m ./output.log
+"""
+
+from io import StringIO
+
+from sqlalchemy import create_engine
+from sqlalchemy.ext.declarative import declarative_base
+from sqlalchemy.orm import sessionmaker
+from sqlalchemy.sql.expression import func #used for sql max() function
+
+import arrow
+import orm_egauge
+import os
+import pandas
+import requests
+import sqlalchemy
+import sys
+import traceback
+
+# used to initialize sqlalchemy ORM
+BASE = declarative_base()
+SCRIPT_NAME = os.path.basename(__file__)
+ERROR_LOG = 'error.log'
+
+def pull_egauge_data(sensor_id='31871', db_url='postgresql:///egauge', unit_of_time='m', output_file='./output.log'):
+    """
+    This function pulls egauge sensor data.
+
+    Data pulled will start at and include the time read from the timestamp log up to (but not including) the current time. The data (not including the headers) pulled from the sensors by this program will be appended to a csv file ('output.log' by default).
+    Any exceptions thrown during execution will be logged to a file.
+
+    Keyword arguments:
+        sensor_id: a string representing the id of the egauge sensor
+        unit_of_time: the unit of time we want the data returned as
+        output_file: the name of the file that egauge sensor data will be appended to
+    """
+
+    # cast sensor_id as int for consistency since id in database is int
+    sensor_id = int(sensor_id)
+    is_success = False
+    current_time = arrow.now().floor('minute')
```

Comment:
Why the floor minute? Won't this approximate the seconds? Why not the exact time?

Reply:
The current_time variable functions as the original source of all success times that will be inserted into and retrieved from the database (barring manually working with the database).

Because the egauge sensors that I've tested sample data at the minute level, I floored current_time to the minute to keep timestamp behavior (especially for api requests and db retrieval/insertion) more consistent and simple.

Do you think that it will be more useful for current_time to include seconds?

I could see it possibly being useful for logging error messages, but I'm not sure if we want to prioritize that over consistency and simplicity in the egauge script.

Reply:
Based on phone call discussion as a reminder:

* Please put a comment on floor to indicate that is an artifact to avoid overlapping readings on line 48.
* We agree on a unit test to be done by a separate issue to test at the same minute and different seconds to evaluate if the reading value is consistently the same (because otherwise we may be trashing away measurements at the second level which would still show a row at the minute level after averaging, just with different number).
* We want to store on the database the reading timestamp that was successful and as well this same timestamp for the last stored reading that is used on the script to select the next time window to download the data. The script will then use this and adjust accordingly (please put a comment indicating why you are adjusting which is to get the immediately next reading), and be careful to not increment if that was a failed insertion!